### ISSUE TITLE
error: path '/nix/store/q0lqpbbxcn2y4f23nhms7ncj7gvwbid1-...' does not exist and cannot be created

### ISSUE THREAD

**Issue by domenkozar** [NixOS/nix#4462]
The error is confusing to the user as they have no idea what's going on.

It should explain what Nix tried to do to create that path and what is possible that it's wrong.

**Comment by zimbatm** [comment]
This happens when a store path is hard-coded in the nix code, isn't it?

User generates some .nix file, which includes a store path. They can import it, life is great. Sometime later, after a garbage-collection, they start getting that error.

**Comment by domenkozar** [comment]
Yeah, it's also possible it can be substituted (that's how cachix install works).

The problem happens when somebody overrides default substituters and they this error (the real fix here is to fix UX of cache.nixos.org as default cache).

**Comment by nh2** [comment]
Potentially related: https://github.com/NixOS/nixpkgs/issues/144628

**Comment by GavinRay97** [comment]
I installed Nix today on Ubuntu 21.10, the first two things that happened:

1. Tried to install `stack` using the command in the docs. It proceeded to try to download and compile +1,000 packages, including low-level stuff like `glibc` and various linux utilities

```ps1
user@MSI:~$ nix-env -f "<nixpkgs>" -iA stack
installing 'stack-2.7.3'

these 698 derivations will be built:
  /nix/store/dhvmh2hb7900z4ki92cfiwap7rhhnqfv-bootstrap-stage0-stdenv-linux.drv
  ...
these 587 paths will be fetched (1125.72 MiB download, 2348.57 MiB unpacked):
  /nix/store/02z0ga1d6p21zkvcwiz8s42qwdzdzl72-cmake-3.21.2.tar.gz
```

2. I tried to install `cachix`, using the command in it's docs:
```ps1
user@MSI:~/projects$ nix-env -iA cachix -f https://cachix.org/api/v1/install

installing 'cachix-0.6.1'
error: path '/nix/store/3lqc09mn31yd05mcy6gcd9nsbcwwphic-cachix-0.6.1' does not exist and cannot be created
```

🙁 I'm not sure if # 1 was intended to happen, and I don't know how to fix # 2 or what's actually happening

**Comment by GavinRay97** [comment]
@domenkozar Yeah it turned out to be exactly that -- I created this PR https://github.com/NixOS/nix/pull/5749 to help prevent idiots like myself from accidentally doing this 😅

**Comment by JosephLucas** [comment]
I stumbled on that issue with a `(import ../../foo.nix)` with no corresponding path in the store. Nix said that its path in the store did not exist or could not be created. The fix was to build foo.nix aside to have it in the store.
